crates/subspace-farmer/src/bench_rpc_client.rs

```diff
@@ -0,0 +1,201 @@
+        let (archived_segments_sender, archived_segments_receiver) = mpsc::channel(10);
+        let (acknowledge_archived_segment_sender, mut acknowledge_archived_segment_receiver) =
+            mpsc::channel(1);
+
+        let slot_info_handler = tokio::spawn(async move {
+            let mut slot_number = 0;
+            let mut next_salt = rand::random();
+            loop {
+                let global_challenge = rand::random();
+                next_salt = {
+                    let (salt, next_salt) = (next_salt, rand::random());
+                    let slot_info = SlotInfo {
+                        slot_number,
+                        global_challenge,
+                        salt: next_salt,
+                        next_salt: salt,
+                        solution_range: rand::random(),
+                    };
+                    if slot_info_sender.send(slot_info).await.is_err() {
+                        break;
+                    };
+                    Some(next_salt)
+                };
+
+                tokio::time::sleep(Duration::from_secs(2 * 60)).await;
+
+                slot_number += 1;
+            }
+        });
+
+        let segment_producer_handle = tokio::spawn(async move {
+            let mut segment_index = 0;
+            let mut last_archived_block = LastArchivedBlock {
+                number: 0,
+                archived_progress: ArchivedBlockProgress::Partial(0),
+            };
+            loop {
+                last_archived_block
+                    .archived_progress
+                    .set_partial(segment_index as u32);
+
+                let archived_segment = {
+                    let root_block = RootBlock::V0 {
+                        segment_index,
+                        records_root: Sha256Hash::default(),
+                        prev_root_block_hash: Sha256Hash::default(),
+                        last_archived_block,
+                    };
+
+                    let mut pieces = FlatPieces::new(100);
+                    rand::thread_rng().fill(pieces.as_mut());
+
+                    let objects = std::iter::repeat_with(|| PieceObject::V0 {
+                        hash: rand::random(),
+                        offset: rand::random(),
+                    })
+                    .take(100)
+                    .collect();
+
+                    ArchivedSegment {
+                        root_block,
+                        pieces,
+                        object_mapping: vec![PieceObjectMapping { objects }],
+                    }
+                };
+
+                if archived_segments_sender
+                    .send(archived_segment)
+                    .await
+                    .is_err()
+                {
+                    break;
+                }
+                if acknowledge_archived_segment_receiver.recv().await.is_none() {
+                    break;
+                }
+
+                segment_index += 1;
+            }
+        });
```

Comment:
Should these be a part of the client itself though? It seems to be that these should be fed into the client from the outside, so you can control it if need be.

Reply:
I can't imagine a good API apart from asking 2 closures from user which would generate segments and slot information. Do you think its fine?

Reply:
I think it'll be better, maybe even two channels, depends on how you implement it